File: web/src/lib/discord.ts
export async function sendDiscordWebhook(
  webhookUrl: string,
  content: string,
): Promise<void> {
  if (!webhookUrl) throw new Error("Missing Discord webhook URL");
  if (!content) throw new Error("Missing content");

  const resp = await fetch(webhookUrl, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ content }),
    cache: "no-store",
  });
  if (!resp.ok) {
    const text = await resp.text().catch(() => "");
    throw new Error(
      `Discord webhook failed ${resp.status}: ${text.slice(0, 300)}`,
    );
  }
}

File: dags/ytb_elt/notify/discord.py
import json
import logging

import requests

logger = logging.getLogger(__name__)


def send_discord_webhook(*, webhook_url: str, content: str) -> None:
    """
    Send a message to a Discord channel via an Incoming Webhook URL.

    Discord expects JSON: {"content": "..."}.
    """
    if not webhook_url:
        raise ValueError("webhook_url is required")
    if not content:
        raise ValueError("content is required")

    payload = {"content": content}
    resp = requests.post(webhook_url, data=json.dumps(payload), headers={"Content-Type": "application/json"}, timeout=15)
    if resp.status_code >= 400:
        logger.error("Discord webhook failed: %s %s", resp.status_code, resp.text[:3000])
        resp.raise_for_status()

